applications/TrilinosApplication/custom_strategies/builder_and_solvers/trilinos_block_builder_and_solver.h

```diff
@@ -4,954 +4,842 @@
+        BuildRHS(pScheme, rModelPart, rb);
+        SystemSolveWithPhysics(rA, rDx, rb, rModelPart);
 
         KRATOS_CATCH("")
     }
 
-    //**************************************************************************
-    //**************************************************************************
-    void BuildRHS(
-        typename TSchemeType::Pointer pScheme,
-        ModelPart& r_model_part,
-        TSystemVectorType& b) override
+    /**
+     * @brief Function to perform the build of the RHS.
+     * @details The vector could be sized as the total number of dofs or as the
+     * number of unrestrained ones
+     * @param pScheme The integration scheme considered
+     * @param rModelPart The model part of the problem to solve
+     */
+    void BuildRHS(typename TSchemeType::Pointer pScheme,
+                  ModelPart& rModelPart,
+                  TSystemVectorType& rb) override
     {
         KRATOS_TRY
 
-        //Getting the Elements
-        ElementsArrayType& pElements = r_model_part.Elements();
+        const int nelements = static_cast<int>(rModelPart.Elements().size());
+        const int nconditions = static_cast<int>(rModelPart.Conditions().size());
+        // Resetting to zero the vector of reactions
+        TSparseSpace::SetToZero(*BaseType::mpReactionsVector); // TODO: Check if required
 
-        //getting the array of the conditions
-        ConditionsArrayType& ConditionsArray = r_model_part.Conditions();
-
-        ProcessInfo& CurrentProcessInfo = r_model_part.GetProcessInfo();
-
-        //resetting to zero the vector of reactions
-        // 			TSparseSpace::SetToZero(BaseType::mReactionsVector);
-
-        //contributions to the system
+        // Contributions to the system
+        LocalSystemMatrixType LHS_Contribution = LocalSystemMatrixType(0, 0);
         LocalSystemVectorType RHS_Contribution = LocalSystemVectorType(0);
 
-        //vector containing the localization in the system of the different
-        //terms
-        Element::EquationIdVectorType EquationId;
+        // vector containing the localization in the system of the different terms
+        Element::EquationIdVectorType equation_ids_vector;
+        ProcessInfo& r_current_process_info = rModelPart.GetProcessInfo();
+        ModelPart::ElementsContainerType::iterator el_begin = rModelPart.ElementsBegin();
+        ModelPart::ConditionsContainerType::iterator cond_begin =
+            rModelPart.ConditionsBegin();
 
         // assemble all elements
-        for (typename ElementsArrayType::ptr_iterator it=pElements.ptr_begin(); it!=pElements.ptr_end(); ++it)
-        {
-            //calculate elemental Right Hand Side Contribution
-            pScheme->Calculate_RHS_Contribution(*it,RHS_Contribution,EquationId,CurrentProcessInfo);
-
-            //assemble the elemental contribution
-            TSparseSpace::AssembleRHS(b,RHS_Contribution,EquationId);
+        for (int k = 0; k < nelements; k++) {
```

Comment:
As in the Build method, I would either add omp support or keep the standard loops. In either case, I would suggest looping on ptr_iterators.